### ISSUE TITLE
Sourcefinder code needs unit tests of Gaussian fitting / data-masking interplay

### ISSUE THREAD

**Issue by timstaley** [transientskp/tkp#428]
I've been trying to wrap my head around the sourcefinder code, in order to sensibly implement #409. 
I'm a bit alarmed to read things like this line:
https://github.com/transientskp/tkp/blob/9d4a2f46b7d000cec08ec2bccfca564cf9bce368/tkp/sourcefinder/extract.py#L664

If I'm understanding correctly, these few lines are attempting to figure out the valid region of data that's about to be fitted. If the valid region is more than 2 pixels across in both dimensions, we go on to fit a Gaussian. Fine. The issue is that we've determined the valid region by checking for **nonzero** rather than **masked**. So it's testing the wrong condition (unless I've missed a trick - second opinions welcome).

We test the basic Gaussian fitting procedure, but we need to verify that the larger 'source_profile_and_measures' routine performs correctly in the presence of masked data.

So: I think we should get this checked (and fixed, if required) for 2.1, even if it means changing (fixing!) the sourcefinder behaviour. Does that sound reasonable? @AntoniaR @ajstewart, would appreciate your opinion on this last bit.

**Comment by timstaley** [comment]
Just thinking through the consequences of this bug. We end up attempting to fit a Gaussian when we shouldn't, e.g. if there's only one or two pixels above the analysis threshold. This might lead to some failed Gaussian fits that should never have been attempted in the first place. Best case is that these fail to converge and we revert back to a sensible moments-only source measurement. Worst case is that it converges to something unreasonable - this might explain some of the weird super-inflated fits we get from time to time when 'force-beam' is switched off?

**Comment by jdswinbank** [comment]
I'm not sure this is as bad as you think it is. Specifically, ``nonzero()`` ignores masked elements. Thus:
```python
In [1]: import numpy

In [2]: a = numpy.ma.array([1,0,1,0], mask=[True, False, False, True])

In [3]: a.nonzero()
Out[3]: (array([2]),)
```
In other words, we're correctly discarding masked pixels, but potentially incorrectly ignoring pixels which are set to zero and *not* masked. Given that the is an array of floats, the chances of that happening seem quite small.

**Comment by timstaley** [comment]
Ah - OK. Thanks for checking that out John - I was conflating it with numpy.nonzero, which of course makes no mention of masked pixels in the docs. So it's probably not doing anything bad - though I might still consider using a clearer / more logically correct function. Or at least adding a comment.

**Comment by timstaley** [comment]
Realistically this won't happen before 3.0, reassigning...

**Comment by AntoniaR** [comment]
@bartscheers is this still an issue with the latest PySE?

**Comment by HannoSpreeuw** [comment]
The required unit tests that @timstaley mentions have been implemented through the [TestMaskedSource unit tests](https://github.com/transientskp/tkp/blob/8a19cd23c7141c66c1ee8e42295957bbcf809531/tests/test_sourcefinder/test_image.py#L378C7-L378C23).

These tests have actually been added by @jdswinbank before this issue was opened, i.e. on March 1, 2013, as per [commit 414a0fb](https://github.com/transientskp/tkp/commit/414a0fb615f50fefbece9b1869b00c572ea340330).

In a broader context, the required width in both dimensions for Gauss fitting, that needed some enhancement in the sense of a more robust criterion, since _a thin strip from the upper left corner to the bottom right corner of the rectangular slice encompassing the island would pass this test, although its width could be as small as a single pixel_. 
That has been addressed by [commit 3ff926e](https://github.com/transientskp/pyse/commit/3ff926e60827c354822d2987e4d6569279e986ab). Should be merged into PySE's `master` branch soon.

Finally, the issue of `weird super-inflated (i.e. large semimajor axis etc) source-fits` is addressed by introducing bounds to gaussian fits. This works remarkably well and has been added by [commit a5f8209](https://github.com/transientskp/pyse/commit/a5f82098292f94c2436928fad9205c7a9dd34c5b), completed by a handful of subsequent commits. To be merged soon as well.

**Comment by AntoniaR** [comment]
Excellent news! Thanks! I will close this issue.
